### PR TITLE
Mock combinators

### PR DESCRIPTION
**Pull request by pcapriotti**
This PR introduces a very simple system for specifying federation mocks in a composable way. This replaces previous attempts (in Galley) that were only adopted in a few tests. All instances of mocked federator are now using the new system.

The system itself is a super-minimalistic "web framework" of sorts, based on a `ReaderT FederatedRequest MaybeT` monad transformer. You use `guard` to expect a certain request, and the `Alternative` combinators to specify multiple responses based on various criteria.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Review by smatting — APPROVED** [review]

File: services/galley/test/integration/API/Util.hs [hunk 1]
```diff
@@ -2772,28 +2766,11 @@ checkTimeout = 3 # Second
...
--- | Only Brig is mocked.
-onlyMockedFederatedBrigResponse :: [(Qualified UserId, Text)] -> FederatedRequest -> Value
-onlyMockedFederatedBrigResponse users =
-  joinMockedFederatedResponses
-    (mockedFederatedBrigResponse users)
-    (const Nothing)
+mockedFederatedBrigResponse :: [(Qualified UserId, Text)] -> Mock LByteString
+mockedFederatedBrigResponse users = do
+  comp <- frComponent <$> getRequest
+  guard (comp == Brig)
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Util.hs:2750]: This is `guardComponent`, right?

File: services/galley/test/integration/API/Util.hs [hunk 2]
```diff
@@ -2572,26 +2573,18 @@ mkProfile quid name =
...
---
--- The `resp :: FederatedRequest -> a` argument can be used to provide a fake
--- federator response (of an arbitrary JSON-serialisable type a) for every
--- expected request.
-withTempMockFederator ::
-  ToJSON a =>
-  (FederatedRequest -> a) ->
-  TestM b ->
-  TestM (b, [FederatedRequest])
-withTempMockFederator resp = withTempMockFederator' $ pure . encode . resp
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Util.hs:2584]: Given that `withTempMockFederator` is gone, how about renaming the remaining `withTempMockFederator'` to `withTempMockFederator`?

File: services/galley/test/integration/API/MLS.hs [hunk 3]
```diff
@@ -2276,3 +2149,38 @@ getGroupInfoDisabled = do
     withMLSDisabled $
       getGroupInfo (qUnqualified alice) qcnv
         !!! assertMLSNotEnabled
+
+--------------------------------------------------------------------------------
+-- Mock utilities
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:2154]: How about moving them to a utilities module?

File: services/galley/test/integration/API/Federation.hs [hunk 4]
```diff
@@ -878,16 +874,15 @@ sendMessage = do
...
-              ( Map.fromList
-                  [ (chadId, Set.singleton (PubClient chadClient Nothing)),
-                    (bobId, Set.singleton (PubClient bobClient Nothing))
-                  ]
-              )
-        | otherwise = toJSON ()
-  (_, requests2) <- withTempMockFederator responses2 $ do
+  let mock = do
+        comp <- frComponent <$> getRequest
+        guard (comp == Brig)
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:879]: This is `guardComponent`.

File: services/galley/test/integration/API/Federation.hs [hunk 5]
```diff
@@ -842,12 +838,12 @@ sendMessage = do
   connectWithRemoteUser aliceId bob
   connectWithRemoteUser aliceId chad
   -- conversation
-  let responses1 req
-        | frComponent req == Brig =
-            toJSON [bobProfile, chadProfile]
-        | otherwise = toJSON ()
+  let responses1 = do
+        comp <- frComponent <$> getRequest
+        guard (comp == Brig)
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/Federation.hs:843]: You could use `guardComponent` instead.

**Review by mdimjasevic — APPROVED** [review]

File: services/galley/test/integration/API/Util.hs (hunk 2 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Util.hs:2584]: It's not gone, it's the low-level function that is used by `withTempMockFederator'`. We could consider condensing all of these into a single function that takes a `Mock` parameter, but I would leave it for another PR.

File: services/galley/test/integration/API/Federation.hs (hunk 4 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs:879]: Fixed.

File: services/galley/test/integration/API/Federation.hs (hunk 5 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/Federation.hs:843]: Fixed

File: services/galley/test/integration/API/MLS.hs (hunk 3 above)
Comment (pcapriotti) [reply services/galley/test/integration/API/MLS.hs]: Done.

**Pull request merged**